### PR TITLE
Hotfix/books availability

### PR DESCRIPTION
**Pull request by mekarpeles**
[WIP] requires:
- replacement of link to @rchrd2's dev availability API w/ production
- round of QA from Jeff & Benjamin
- updating the popular + waitlist carousel tests

### CODE REVIEW

File: openlibrary/plugins/openlibrary/home.py [hunk 1]
```diff
@@ -214,7 +213,7 @@ def get_returncart(limit):
 get_returncart = cache.memcache_memoize(get_returncart, "home.get_returncart", timeout=60)
 
 @public
-def readonline_carousel(id="read-carousel"):
+def readonline_carousel(id="CarouselClassics"):
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/home.py]: Same comment re: titlecase as above.

File: openlibrary/plugins/openlibrary/js/availability.js [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+$(function(){
+
+    getEditions = function(olids, callback, decoration, cssid) {
+        var url = '/api/editions?olids=' + olids.join(',');
+        if (decoration) {
+            url += '&decoration=' + decoration;
+        }
+	if (cssid) {
+	    url += '&cssid=' + cssid;
+	}
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/availability.js]: Indenting is one space short on this `if` clause

File: openlibrary/plugins/openlibrary/js/availability.js [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+     */
+    updateBookAvailability = function(selector) {
+        selector = selector || ''; // effectively sets default value of selector
+        selector += '[data-ocaid]';
+	var books = {};  // lets us keep track of which ocaids came from
+                         // which book (i.e. edition or work). As we learn
+                         // ocaids are available, we'll need a way to
+                         // determine which edition or work this ocaid
+                         // comes from.
+	var ocaids = []  // a full set of ocaids spanning all books
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/availability.js]: Missing semicolon

File: openlibrary/plugins/openlibrary/js/availability.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+     * Finds DOM elements for borrowable books (i.e. they have a
+     * data-ocaid field and a data-key field) within the specified
+     * scope of `selector` and updates they displayed borrow status
+     * and ebook links to reflect correct statuses and available
+     * copies.
+     */
+    updateBookAvailability = function(selector) {
+        selector = selector || ''; // effectively sets default value of selector
+        selector += '[data-ocaid]';
+	var books = {};  // lets us keep track of which ocaids came from
```
Comment (bfalling) [inline openlibrary/plugins/openlibrary/js/availability.js]: Check indentation on this line and all below

File: openlibrary/core/lending.py [hunk 5]
```diff
@@ -64,22 +64,30 @@ def setup(config):
...
+    params:
+        identifiers (list) - ocaids; Internet Archive item identifiers
+        acs (int) - do a check to mark acs loans as unavailable
+        restricted (int) - flag book is apart of a restricted collection
+
     """
-    url = AVAILABILITY_API + '?action=availability'
+    url = (AVAILABILITY_API + '?action=availability&exact=%s&validate=%s'
+           % (acs, restricted))
+    print(url)
```
Comment (bfalling) [inline openlibrary/core/lending.py]: Prefer using logger over `print()`, I think

File: openlibrary/plugins/upstream/borrow.py [hunk 6]
```diff
@@ -74,6 +74,30 @@
...
 # Handler for /books/{bookid}/{title}/borrow
+class checkout_with_ocaid(delegate.page):
+
+    path = "/borrow/ia/(.*)"
+
+    def GET(self, ocaid):
+        """Redirect shim: Translate an IA identifier into an OL identifier and
+        then redirects user to the canonical OL borrow page.
+        """
+        print(ocaid)
```
Comment (bfalling) [inline openlibrary/plugins/upstream/borrow.py]: Again, prefer logging over `print`ing, I think, here and below

File: openlibrary/templates/home/index.html [hunk 7]
```diff
@@ -37,3 +37,149 @@
...
+                xhr.setRequestHeader("Content-Type", "application/json");
+                xhr.setRequestHeader("Accept", "application/json");
+            },
+            success: function(result) {
+                return callback(result);
+            }
+        });
+    };
+
+    var recGetPopularEditions = function(all_books, available, waitlist, num_slots, start, limit, callback) {
```
Comment (bfalling) [inline openlibrary/templates/home/index.html]: What is "rec" prefix? Maybe just spell it out?

File: openlibrary/templates/home/index.html [hunk 8]
```diff
@@ -37,3 +37,149 @@
...
+        \$('#' + cssid + ' li').each(function(index, elem) {
+            \$(elem).find(".coverMagic").removeClass("hiddenCoverMagic");
+            var cover = \$(elem).find("img.cover");
+            if (cover.length > 0) {
+                cover.attr("src", cover.attr("xsrc"));
+            }
+        });
+
+        \$("#" + cssid).jcarousel({
+            scroll:6,
```
Comment (bfalling) [inline openlibrary/templates/home/index.html]: Explain magic constant 6 or parameterize

File: openlibrary/templates/home/popular.html [hunk 9]
```diff
@@ -7,4 +7,4 @@ <h2 class="inline"><a href="/people/openlibrary/lists/OL104411L/Popular">Popular
   </p>
 </div>
 
-$:render_template("books/carousel", books or [], id="CarouselPopular")
+$:render_template("books/carousel_shell", books or [], id="CarouselPopular")
```
Comment (bfalling) [inline openlibrary/templates/home/popular.html]: Same comment re: titlecase. Ditto in waitlist.html below.

File: openlibrary/templates/lib/covers.html [hunk 10]
```diff
@@ -27,6 +27,9 @@
     carousel.add(index, "<h3>Loading books...</h3>");
     page.loadPage(index-1, function(data) {
```
Comment (bfalling) [inline openlibrary/templates/lib/covers.html]: Spaces around operator?

File: openlibrary/templates/lib/covers.html [hunk 11]
```diff
@@ -27,6 +27,9 @@
...
         carousel.add(index, page.renderCovers(data));
+
+        console.log('availability about to get called...');
+        updateBookAvailability('#coversCarousel li);
```
Comment (bfalling) [inline openlibrary/templates/lib/covers.html]: Syntax error? Missing closing single-quote? (GitHub shows this line and following code with warning red syntax coloring...)

File: openlibrary/core/lending.py [hunk 12]
```diff
@@ -64,22 +64,30 @@ def setup(config):
     config_ia_ol_shared_key = config.get('ia_ol_shared_key')
 
 
-def is_borrowable(identifiers):
+def is_borrowable(identifiers, acs=0, restricted=0):
     """Takes a list of archive.org ocaids and returns json indicating
```
Comment (mekarpeles) [inline openlibrary/core/lending.py]: addressed in 71f0f8be13237ae4f4b7884c66af1039b6c86ecc

File: openlibrary/core/lending.py (hunk 5 above)
Comment (mekarpeles) [inline openlibrary/core/lending.py]: addressed in 71f0f8be13237ae4f4b7884c66af1039b6c86ecc

File: openlibrary/plugins/openlibrary/api.py [hunk 13]
```diff
@@ -3,12 +3,104 @@
...
+
+    def GET(self):
+        """Returns `limit` popular book tuples of form [ocaid, olid] starting
+        at `start`
+
+        Should be cached through memcached.
+        """
+        i = web.input(start='0', limit='100')
+        start, limit = int(i.start), int(i.limit),
+        books = popular_editions[start:start+limit]
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py]: Following https://www.python.org/dev/peps/pep-0008/#id26, adding spacing though not required around '+'

File: openlibrary/plugins/openlibrary/api.py [hunk 14]
```diff
@@ -3,12 +3,104 @@
...
+            'limit': limit,
+            'start': start,
+            'next': start + limit
+        } if books else {'error': 'out_of_range'}
+        return delegate.RawText(simplejson.dumps(result),
+                                content_type="application/json")
+
+def format_edition(edition):
+    """This should be moved to a books or carousel model"""
+    from openlibrary.core import ia
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py]: It's good practice though there are two major exceptions. (1) importing within the lexical scope of a function can help prevent circular import dependencies and (2) and prevent load overhead if only one method needs the contents of a library. In this case it should be negligible; approving your recommendation

File: openlibrary/plugins/openlibrary/api.py [hunk 15]
```diff
@@ -3,12 +3,104 @@
...
+    book = {
+        'ocaid': edition.get('ocaid'),
+        'title': edition.title or None,
+        'key': edition.key,
+        'url': edition.url(),
+        'authors': [web.storage(key=a.key, name=a.name or None)
+                    for a in edition.get_authors()],
+        'collections': collections,
+        'protected': not any([c not in collections for c in
+                              ['lendinglibrary', 'browserlending', 'inlibrary']])
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py]: We should be able to simplify this to:

    [c in collections for c in ['lendinglibrary', 'browserlending', 'inlibrary']]

File: openlibrary/plugins/openlibrary/api.py [hunk 16]
```diff
@@ -3,12 +3,104 @@
...
+        i = web.input(olids='', decoration=None, cssid=None)
+        keys = i.olids.split(',')
+
+        if i.decoration == "carousel_item":
+            decorate = lambda book: render_template(
+                'books/carousel_item', web.storage(book), id=i.cssid).__body__
+        else:
+            decorate = lambda book: book  # identity
+
+        if len(keys) > 100:
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py]: changing to max_limit (as defined in the function signature)

File: openlibrary/plugins/openlibrary/api.py [hunk 17]
```diff
@@ -3,12 +3,104 @@
...
+            'start': start,
+            'next': start + limit
+        } if books else {'error': 'out_of_range'}
+        return delegate.RawText(simplejson.dumps(result),
+                                content_type="application/json")
+
+def format_edition(edition):
+    """This should be moved to a books or carousel model"""
+    from openlibrary.core import ia
+    collections = ia.get_meta_xml(edition.get('ocaid')).get("collection", [])
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py]: Yes, this pre-existing API is completely confusing. It is the archive.org json metadata api which used to be xml.

File: openlibrary/plugins/openlibrary/home.py (hunk 1 above)
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/home.py]: Addressed in 989de3339002515b4ae5238ae5d95b33e6ffc711

File: openlibrary/plugins/openlibrary/home.py [hunk 18]
```diff
@@ -198,7 +197,7 @@ def render_returncart(limit=60, randomize=True):
     if randomize:
         random.shuffle(data)
     data = data[:limit]
-    return render_template("books/carousel", storify(data), id="returncart_carousel")
+    return render_template("books/carousel", storify(data), id="CarouselReturns")
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/home.py]: Addressed in 989de3339002515b4ae5238ae5d95b33e6ffc711

File: openlibrary/plugins/upstream/borrow.py (hunk 6 above)
Comment (mekarpeles) [inline openlibrary/plugins/upstream/borrow.py]: Removed (for debugging only)

File: openlibrary/templates/home/index.html (hunk 7 above)
Comment (mekarpeles) [inline openlibrary/templates/home/index.html]: recursively (changed)

File: openlibrary/templates/home/index.html (hunk 8 above)
Comment (mekarpeles) [inline openlibrary/templates/home/index.html]: This is the constructor to jcarousel, it specifies scrolling should be done 6 books at a time (e.g. a page at a time). Added comment to code

File: openlibrary/templates/home/popular.html (hunk 9 above)
Comment (mekarpeles) [inline openlibrary/templates/home/popular.html]: addressed

File: openlibrary/plugins/upstream/code.py [hunk 19]
```diff
@@ -27,12 +27,13 @@
...
 
 if not config.get('coverstore_url'):
-    config.coverstore_url = "http://covers.openlibrary.org"
+    config.coverstore_url = "https://covers.openlibrary.org"
 
 class static(delegate.page):
     path = "/images/.*"
     def GET(self):
-        raise web.seeother('/static/' + web.ctx.path)
+        host = 'https://%s' % web.ctx.host if 'openlibrary.org' in web.ctx.host else ''
```
Comment (bfalling) [inline openlibrary/plugins/upstream/code.py]: @mekarpeles: What's your motivation behind explicitly coding in the HTTPS host? Will not the relative path `/static/` already use the existing protocol (which should be HTTPS)?

Comment (mekarpeles) [inline openlibrary/plugins/upstream/code.py]: This uses the gunicorn web.ctx var which under the hood appears to be http (resulting in insecure http request attempts for images)

**Pull request merged**